infra/build/functions/ci.py

```diff
@@ -0,0 +1,260 @@
+          build_project.get_project_data(project_name))
+    except FileNotFoundError:
+      logging.error('Couldn\'t get project data. Skipping %s.', project_name)
+      continue
+
+    project_yaml['sanitizers'] = list(
+        set(project_yaml['sanitizers']).intersection(set(sanitizers)))
+
+    project_yaml['fuzzing_engines'] = list(
+        set(project_yaml['fuzzing_engines']).intersection(
+            set(args.fuzzing_engines)))
+
+    if not project_yaml['sanitizers'] or not project_yaml['fuzzing_engines']:
+      logging.info('Nothing to build for this project: %s.', project_name)
+      continue
+
+    steps = get_build_steps_func(project_name, project_yaml,
+                                 dockerfile_contents, IMAGE_PROJECT,
+                                 BASE_IMAGES_PROJECT, config)
+    if not steps:
+      logging.error('No steps. Skipping %s.', project_name)
+      continue
+
+    project_statuses[project_name].set_build_id(
+        build_project.run_build(project_name, steps, credentials, build_type),
+        build_type)
+
+  return project_statuses
+
+
+def get_build_status_from_gcb(cloudbuild_api, build_id):
+  """Returns the status of the build: |build_id| from cloudbuild_api."""
+  build_result = cloudbuild_api.get(projectId=IMAGE_PROJECT,
+                                    id=build_id).execute()
+  return build_result['status']
+
+
+def check_finished(project_status, build_type, cloudbuild_api):
+  """Checks that the |build_type| build is complete. Updates |project_status| if
+  complete."""
+  build_id = project_status.build_id[build_type]
+  if not build_id:
+    return
+  print('build_id', build_id)
+  build_status = get_build_status_from_gcb(cloudbuild_api, build_id)
+  if build_status not in FINISHED_BUILD_STATUSES:
+    return
+  project_status.set_build_result(build_status == 'SUCCESS', build_type)
+
+
+def wait_on_builds(project_statuses, credentials):
+  """Waits on |builds|. Returns True if all builds succeed."""
+  results = collections.defaultdict(lambda: {
+      'coverage': False,
+      'fuzzing': False
+  })
+
+  cloudbuild = cloud_build('cloudbuild',
+                           'v1',
+                           credentials=credentials,
+                           cache_discovery=False)
+  cloudbuild_api = cloudbuild.projects().builds()  # pylint: disable=no-member
+
+  wait_projects = project_statuses.copy()
+  while wait_projects:
+    logging.info('Polling')
+    for project, project_status in list(wait_projects.items()):
+      check_finished(project_status, 'fuzzing', cloudbuild_api)
+      check_finished(project_status, 'coverage', cloudbuild_api)
+      if all(project_status.build_finished.values()):
+        del wait_projects[project]
+
+  print('Printing results')
+  print('Project, Statuses')
+  for project, project_status in project_statuses.items():
+    print(project, project_status.build_result)
+
+  return all(all(statuses) for statuses in results.values())
```

Comment:
To account for legitimate flake, we probably want some kind of threshold instead right?

Reply:
Absolutely, maybe we can leave that for later?

Reply:
SGTM